sw/device/silicon_creator/rom_ext/rom_ext.c

```diff
@@ -186,7 +187,14 @@ static rom_error_t rom_ext_try_boot(void) {
 void rom_ext_main(void) {
   rom_ext_init();
   OT_DISCARD(rom_printf("Starting ROM_EXT\r\n"));
-  shutdown_finalize(rom_ext_try_boot());
+  rom_error_t error = rom_ext_try_boot();
+  // If the boot failed, check whether we should enter bootstrap mode.
+  if (error != kErrorOk && rom_ext_bootstrap_enabled() == kHardenedBoolTrue) {
+    HARDENED_CHECK_EQ(rom_ext_bootstrap_enabled(), kHardenedBoolTrue);
+    OT_DISCARD(rom_printf("Entering ROM_EXT bootstrap\r\n"));
```

Comment:
So this printf is not strictly necessary, but I figured it simplifies testing `rom_ext_bootstrap_enabled()` via a functest, since we can set the `exit_success` to `"Entering ROM_EXT bootstrap\r\n"`.

Deleting is fine as well — e2e tests can just test the observable behavior, such as whether the requested pages were written to flash.

Reply:
> So this printf is not strictly necessary, but I figured it simplifies testing `rom_ext_bootstrap_enabled()` via a functest, since we can set the `exit_success` to `"Entering ROM_EXT bootstrap\r\n"`.

I see.

> Deleting is fine as well — e2e tests can just test the observable behavior, such as whether the requested pages were written to flash.

I think we can also send jedec_id or read_sfdp commands and check the response.

If you prefer to include this please feel free to add a fpga-specific function implementation in `device_fpga_cw310.c`. You might also want to convert the existing function to a generic `fpga_rom_printf()`. But please note that if you use it in tests those tests would pass only on FPGA.